### PR TITLE
Add --run-mode and update flops_counter

### PR DESCRIPTION
**Pull request by colizz**
## Add --run-mode

Specify the run mode. Syntax: `--run-mode train,val,test`

## Small fix on the floating-point precision issue in --fetch-step

The floating-point error may accumulate and cause e.g. `self.ipos` = 0.99999999 such that the stop point is not triggered

## Small fix on the self-mapping removal code
https://github.com/hqucms/weaver-core/pull/18/commits/eaace4d5dcf0cc7058607c4f771384de12dc592a

## Update flops_counter.

The `flops_counter.py` is updated to the latest version in https://github.com/sovrasov/flops-counter.pytorch
The code is adapted to weaver-core according to its original form

Several changes are observed:
 - FLOPs produced from torch.nn functionals (such as F.relu) and torch operations (such as `torch.bmm`, `torch.mul`) is now considered.
 - more modules are supported to compute FLOPs, e.g. `nn.LayerNorm`, `nn.GELU`.
 - a bug in computing FLOPs in `nn.Linear` on the bias term is fixed in the latest version of the code.

Comparison of the old/new impl. when performed on the latest ParT model.

```diff
1,2c1,2
< [2024-08-06 00:29:52,079] INFO: ParticleTransformerWrapper(
<   |2.123 M, 99.021% Params, 247.353 MMac, 100.000% MACs|
---
> [2024-08-06 00:29:33,592] INFO: ParticleTransformerWrapper(
>   |2.142 M, 99.931% Params, 251.542 MMac, 92.948% MACs|
4c4
<     |2.123 M, 99.021% Params, 247.353 MMac, 100.000% MACs|
---
>     |2.142 M, 99.931% Params, 251.542 MMac, 92.948% MACs|
7c7
<       |0.134 M, 6.254% Params, 17.061 MMac, 6.897% MACs|
---
>       |0.135 M, 6.315% Params, 17.341 MMac, 6.408% MACs|
10,19c10,19
<         |0.134 M, 6.252% Params, 17.057 MMac, 6.896% MACs|
<         (0): LayerNorm((17,), eps=1e-05, elementwise_affine=True, |0.0 M, 0.000% Params, 0.0 Mac, 0.000% MACs|)
<         (1): Linear(in_features=17, out_features=128, bias=True, |0.002 M, 0.107% Params, 278.656 KMac, 0.113% MACs|)
<         (2): GELU(approximate='none', |0.0 M, 0.000% Params, 0.0 Mac, 0.000% MACs|)
<         (3): LayerNorm((128,), eps=1e-05, elementwise_affine=True, |0.0 M, 0.000% Params, 0.0 Mac, 0.000% MACs|)
<         (4): Linear(in_features=128, out_features=512, bias=True, |0.066 M, 3.081% Params, 8.389 MMac, 3.392% MACs|)
<         (5): GELU(approximate='none', |0.0 M, 0.000% Params, 0.0 Mac, 0.000% MACs|)
<         (6): LayerNorm((512,), eps=1e-05, elementwise_affine=True, |0.0 M, 0.000% Params, 0.0 Mac, 0.000% MACs|)
<         (7): Linear(in_features=512, out_features=128, bias=True, |0.066 M, 3.063% Params, 8.389 MMac, 3.391% MACs|)
<         (8): GELU(approximate='none', |0.0 M, 0.000% Params, 0.0 Mac, 0.000% MACs|)
---
>         |0.135 M, 6.314% Params, 17.336 MMac, 6.406% MACs|
>         (0): LayerNorm((17,), eps=1e-05, elementwise_affine=True, |0.0 M, 0.002% Params, 2.176 KMac, 0.001% MACs|)
>         (1): Linear(in_features=17, out_features=128, bias=True, |0.002 M, 0.107% Params, 294.912 KMac, 0.109% MACs|)
>         (2): GELU(approximate='none', |0.0 M, 0.000% Params, 16.384 KMac, 0.006% MACs|)
>         (3): LayerNorm((128,), eps=1e-05, elementwise_affine=True, |0.0 M, 0.012% Params, 16.384 KMac, 0.006% MACs|)
>         (4): Linear(in_features=128, out_features=512, bias=True, |0.066 M, 3.081% Params, 8.454 MMac, 3.124% MACs|)
>         (5): GELU(approximate='none', |0.0 M, 0.000% Params, 65.536 KMac, 0.024% MACs|)
>         (6): LayerNorm((512,), eps=1e-05, elementwise_affine=True, |0.001 M, 0.048% Params, 65.536 KMac, 0.024% MACs|)
>         (7): Linear(in_features=512, out_features=128, bias=True, |0.066 M, 3.063% Params, 8.405 MMac, 3.106% MACs|)
>         (8): GELU(approximate='none', |0.0 M, 0.000% Params, 16.384 KMac, 0.006% MACs|)
23c23
<       |0.01 M, 0.446% Params, 78.993 MMac, 31.935% MACs|
---
>       |0.01 M, 0.446% Params, 80.645 MMac, 29.799% MACs|
25,38c25,38
<         |0.01 M, 0.446% Params, 78.993 MMac, 31.935% MACs|
<         (0): BatchNorm1d(4, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True, |0.0 M, 0.000% Params, 66.048 KMac, 0.027% MACs|)
<         (1): Conv1d(4, 64, kernel_size=(1,), stride=(1,), |0.0 M, 0.015% Params, 2.642 MMac, 1.068% MACs|)
<         (2): BatchNorm1d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True, |0.0 M, 0.006% Params, 1.057 MMac, 0.427% MACs|)
<         (3): GELU(approximate='none', |0.0 M, 0.000% Params, 0.0 Mac, 0.000% MACs|)
<         (4): Conv1d(64, 64, kernel_size=(1,), stride=(1,), |0.004 M, 0.194% Params, 34.345 MMac, 13.885% MACs|)
<         (5): BatchNorm1d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True, |0.0 M, 0.006% Params, 1.057 MMac, 0.427% MACs|)
<         (6): GELU(approximate='none', |0.0 M, 0.000% Params, 0.0 Mac, 0.000% MACs|)
<         (7): Conv1d(64, 64, kernel_size=(1,), stride=(1,), |0.004 M, 0.194% Params, 34.345 MMac, 13.885% MACs|)
<         (8): BatchNorm1d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True, |0.0 M, 0.006% Params, 1.057 MMac, 0.427% MACs|)
<         (9): GELU(approximate='none', |0.0 M, 0.000% Params, 0.0 Mac, 0.000% MACs|)
<         (10): Conv1d(64, 8, kernel_size=(1,), stride=(1,), |0.001 M, 0.024% Params, 4.293 MMac, 1.736% MACs|)
<         (11): BatchNorm1d(8, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True, |0.0 M, 0.001% Params, 132.096 KMac, 0.053% MACs|)
<         (12): GELU(approximate='none', |0.0 M, 0.000% Params, 0.0 Mac, 0.000% MACs|)
---
>         |0.01 M, 0.446% Params, 80.645 MMac, 29.799% MACs|
>         (0): BatchNorm1d(4, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True, |0.0 M, 0.000% Params, 66.048 KMac, 0.024% MACs|)
>         (1): Conv1d(4, 64, kernel_size=(1,), stride=(1,), |0.0 M, 0.015% Params, 2.642 MMac, 0.976% MACs|)
>         (2): BatchNorm1d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True, |0.0 M, 0.006% Params, 1.057 MMac, 0.390% MACs|)
>         (3): GELU(approximate='none', |0.0 M, 0.000% Params, 528.384 KMac, 0.195% MACs|)
>         (4): Conv1d(64, 64, kernel_size=(1,), stride=(1,), |0.004 M, 0.194% Params, 34.345 MMac, 12.691% MACs|)
>         (5): BatchNorm1d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True, |0.0 M, 0.006% Params, 1.057 MMac, 0.390% MACs|)
>         (6): GELU(approximate='none', |0.0 M, 0.000% Params, 528.384 KMac, 0.195% MACs|)
>         (7): Conv1d(64, 64, kernel_size=(1,), stride=(1,), |0.004 M, 0.194% Params, 34.345 MMac, 12.691% MACs|)
>         (8): BatchNorm1d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True, |0.0 M, 0.006% Params, 1.057 MMac, 0.390% MACs|)
>         (9): GELU(approximate='none', |0.0 M, 0.000% Params, 528.384 KMac, 0.195% MACs|)
>         (10): Conv1d(64, 8, kernel_size=(1,), stride=(1,), |0.001 M, 0.024% Params, 4.293 MMac, 1.586% MACs|)
>         (11): BatchNorm1d(8, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True, |0.0 M, 0.001% Params, 132.096 KMac, 0.049% MACs|)
>         (12): GELU(approximate='none', |0.0 M, 0.000% Params, 66.048 KMac, 0.024% MACs|)
43,44c43,44
<         |0.198 M, 9.226% Params, 18.875 MMac, 7.631% MACs|
<         (pre_attn_norm): LayerNorm((128,), eps=1e-05, elementwise_affine=True, |0.0 M, 0.000% Params, 0.0 Mac, 0.000% MACs|)
---
>         |0.2 M, 9.310% Params, 19.153 MMac, 7.077% MACs|
>         (pre_attn_norm): LayerNorm((128,), eps=1e-05, elementwise_affine=True, |0.0 M, 0.012% Params, 16.384 KMac, 0.006% MACs|)
46c46
<           |0.066 M, 3.081% Params, 2.097 MMac, 0.848% MACs|
---
>           |0.066 M, 3.081% Params, 2.114 MMac, 0.781% MACs|
48c48
<           (out_proj): Linear(in_features=128, out_features=128, bias=True, |0.017 M, 0.770% Params, 2.097 MMac, 0.848% MACs|)
---
>           (out_proj): Linear(in_features=128, out_features=128, bias=True, |0.017 M, 0.770% Params, 2.114 MMac, 0.781% MACs|)
50c50
<         (post_attn_norm): LayerNorm((128,), eps=1e-05, elementwise_affine=True, |0.0 M, 0.000% Params, 0.0 Mac, 0.000% MACs|)
---
>         (post_attn_norm): LayerNorm((128,), eps=1e-05, elementwise_affine=True, |0.0 M, 0.012% Params, 16.384 KMac, 0.006% MACs|)
54,56c54,56
<         (pre_fc_norm): LayerNorm((128,), eps=1e-05, elementwise_affine=True, |0.0 M, 0.000% Params, 0.0 Mac, 0.000% MACs|)
<         (fc1): Linear(in_features=128, out_features=512, bias=True, |0.066 M, 3.081% Params, 8.389 MMac, 3.392% MACs|)
<         (act): GELU(approximate='none', |0.0 M, 0.000% Params, 0.0 Mac, 0.000% MACs|)
---
>         (pre_fc_norm): LayerNorm((128,), eps=1e-05, elementwise_affine=True, |0.0 M, 0.012% Params, 16.384 KMac, 0.006% MACs|)
>         (fc1): Linear(in_features=128, out_features=512, bias=True, |0.066 M, 3.081% Params, 8.454 MMac, 3.124% MACs|)
>         (act): GELU(approximate='none', |0.0 M, 0.000% Params, 65.536 KMac, 0.024% MACs|)
58,59c58,59
<         (post_fc_norm): LayerNorm((512,), eps=1e-05, elementwise_affine=True, |0.0 M, 0.000% Params, 0.0 Mac, 0.000% MACs|)
<         (fc2): Linear(in_features=512, out_features=128, bias=True, |0.066 M, 3.063% Params, 8.389 MMac, 3.391% MACs|)
---
>         (post_fc_norm): LayerNorm((512,), eps=1e-05, elementwise_affine=True, |0.001 M, 0.048% Params, 65.536 KMac, 0.024% MACs|)
>         (fc2): Linear(in_features=512, out_features=128, bias=True, |0.066 M, 3.063% Params, 8.405 MMac, 3.106% MACs|)
66,67c66,67
<         |0.198 M, 9.226% Params, 148.224 KMac, 0.060% MACs|
<         (pre_attn_norm): LayerNorm((128,), eps=1e-05, elementwise_affine=True, |0.0 M, 0.000% Params, 0.0 Mac, 0.000% MACs|)
---
>         |0.2 M, 9.310% Params, 166.016 KMac, 0.061% MACs|
>         (pre_attn_norm): LayerNorm((128,), eps=1e-05, elementwise_affine=True, |0.0 M, 0.012% Params, 16.512 KMac, 0.006% MACs|)
69c69
<           |0.066 M, 3.081% Params, 16.512 KMac, 0.007% MACs|
---
>           |0.066 M, 3.081% Params, 16.512 KMac, 0.006% MACs|
71c71
<           (out_proj): Linear(in_features=128, out_features=128, bias=True, |0.017 M, 0.770% Params, 16.512 KMac, 0.007% MACs|)
---
>           (out_proj): Linear(in_features=128, out_features=128, bias=True, |0.017 M, 0.770% Params, 16.512 KMac, 0.006% MACs|)
73c73
<         (post_attn_norm): LayerNorm((128,), eps=1e-05, elementwise_affine=True, |0.0 M, 0.000% Params, 0.0 Mac, 0.000% MACs|)
---
>         (post_attn_norm): LayerNorm((128,), eps=1e-05, elementwise_affine=True, |0.0 M, 0.012% Params, 128.0 Mac, 0.000% MACs|)
77,79c77,79
<         (pre_fc_norm): LayerNorm((128,), eps=1e-05, elementwise_affine=True, |0.0 M, 0.000% Params, 0.0 Mac, 0.000% MACs|)
<         (fc1): Linear(in_features=128, out_features=512, bias=True, |0.066 M, 3.081% Params, 66.048 KMac, 0.027% MACs|)
<         (act): GELU(approximate='none', |0.0 M, 0.000% Params, 0.0 Mac, 0.000% MACs|)
---
>         (pre_fc_norm): LayerNorm((128,), eps=1e-05, elementwise_affine=True, |0.0 M, 0.012% Params, 128.0 Mac, 0.000% MACs|)
>         (fc1): Linear(in_features=128, out_features=512, bias=True, |0.066 M, 3.081% Params, 66.048 KMac, 0.024% MACs|)
>         (act): GELU(approximate='none', |0.0 M, 0.000% Params, 512.0 Mac, 0.000% MACs|)
81,82c81,82
<         (post_fc_norm): LayerNorm((512,), eps=1e-05, elementwise_affine=True, |0.0 M, 0.000% Params, 0.0 Mac, 0.000% MACs|)
<         (fc2): Linear(in_features=512, out_features=128, bias=True, |0.066 M, 3.063% Params, 65.664 KMac, 0.027% MACs|)
---
>         (post_fc_norm): LayerNorm((512,), eps=1e-05, elementwise_affine=True, |0.001 M, 0.048% Params, 512.0 Mac, 0.000% MACs|)
>         (fc2): Linear(in_features=512, out_features=128, bias=True, |0.066 M, 3.063% Params, 65.664 KMac, 0.024% MACs|)
87c87
<     (norm): LayerNorm((128,), eps=1e-05, elementwise_affine=True, |0.0 M, 0.000% Params, 0.0 Mac, 0.000% MACs|)
---
>     (norm): LayerNorm((128,), eps=1e-05, elementwise_affine=True, |0.0 M, 0.012% Params, 128.0 Mac, 0.000% MACs|)
89,90c89,90
<       |0.001 M, 0.060% Params, 1.29 KMac, 0.001% MACs|
<       (0): Linear(in_features=128, out_features=10, bias=True, |0.001 M, 0.060% Params, 1.29 KMac, 0.001% MACs|)
---
>       |0.001 M, 0.060% Params, 1.29 KMac, 0.000% MACs|
>       (0): Linear(in_features=128, out_features=10, bias=True, |0.001 M, 0.060% Params, 1.29 KMac, 0.000% MACs|)
94,95c94,95
< [2024-08-06 00:29:52,081] INFO: Computational complexity:       247.35 MMac
< [2024-08-06 00:29:52,082] INFO: Number of parameters:           2.14 M
---
> [2024-08-06 00:29:33,629] INFO: Computational complexity:       270.63 MMac
> [2024-08-06 00:29:33,629] INFO: Number of parameters:           2.14 M
```

### CODE REVIEW

File: weaver/train.py [hunk 1]
```diff
@@ -19,6 +19,9 @@
 from weaver.utils.data.eval_utils import _register_funcs
 
 parser = argparse.ArgumentParser()
+parser.add_argument('--run-mode', type=str, default='default',
+                    choices=['default', 'train-only', 'val-only', 'test-only'],
+                    help='run mode')
 parser.add_argument('--regression-mode', action='store_true', default=False,
```
Comment (hqucms) [inline weaver/train.py]:
```suggestion
parser.add_argument('--run-mode', type=functools.partial(str.split, sep=','), default='train,val,test',
                    help='comma-separated list of the steps (train|val|test) to run, e.g., `train,test`')
```

File: weaver/train.py [hunk 2]
```diff
@@ -872,37 +875,42 @@ def _main(args):
...
-                best_valid_metric = valid_metric
-                if args.model_prefix and (args.backend is None or local_rank == 0):
-                    shutil.copy2(args.model_prefix + '_epoch-%d_state.pt' %
-                                 epoch, args.model_prefix + '_best_epoch_state.pt')
-                    # torch.save(model, args.model_prefix + '_best_epoch_full.pt')
-            _logger.info('Epoch #%d: Current validation metric: %.5f (best: %.5f)' %
-                         (epoch, valid_metric, best_valid_metric), color='bold')
 
-    if args.data_test:
+            if args.run_mode in ['default', 'train-only']:
```
Comment (hqucms) [inline weaver/train.py]:
```suggestion
            if 'train' in args.run_mode:
```

File: weaver/train.py [hunk 3]
```diff
@@ -872,37 +875,42 @@ def _main(args):
...
+                        os.makedirs(dirname)
+                    prev_ckpts = glob.glob(args.model_prefix + '_epoch-*.pt') if args.auto_clean else []
+                    state_dict = model.module.state_dict() if isinstance(
+                        model, (torch.nn.DataParallel, torch.nn.parallel.DistributedDataParallel)) else model.state_dict()
+                    torch.save(state_dict, args.model_prefix + '_epoch-%d_state.pt' % epoch)
+                    torch.save(opt.state_dict(), args.model_prefix + '_epoch-%d_optimizer.pt' % epoch)
+                    for f in prev_ckpts:
+                        os.remove(f)
+
+            if args.run_mode in ['default', 'val-only']:
```
Comment (hqucms) [inline weaver/train.py]:
```suggestion
            if 'val' in args.run_mode:
```

File: weaver/train.py [hunk 4]
```diff
@@ -872,37 +875,42 @@ def _main(args):
...
+                _logger.info('Epoch #%d validating' % epoch)
+                if args.run_mode == 'val-only':
```
Comment (hqucms) [inline weaver/train.py]:
```suggestion
                if 'train' not in args.run_mode:
```

File: weaver/train.py [hunk 5]
```diff
@@ -872,37 +875,42 @@ def _main(args):
...
+                if is_best_epoch:
+                    best_valid_metric = valid_metric
+                    if args.model_prefix and (args.backend is None or local_rank == 0):
+                        shutil.copy2(args.model_prefix + '_epoch-%d_state.pt' %
+                                    epoch, args.model_prefix + '_best_epoch_state.pt')
+                        # torch.save(model, args.model_prefix + '_best_epoch_full.pt')
+                _logger.info('Epoch #%d: Current validation metric: %.5f (best: %.5f)' %
+                            (epoch, valid_metric, best_valid_metric), color='bold')
+
+    if args.run_mode in ['default', 'test-only'] and args.data_test:
```
Comment (hqucms) [inline weaver/train.py]:
```suggestion
    if 'test' in args.run_mode and args.data_test:
```

**Comment by hqucms** [comment]
Thanks a lot @colizz !
I tweaked a bit the `--run-mode` to make it more flexible. Let me know what you think.

File: weaver/train.py (hunk 1 above)
Comment (colizz) [reply weaver/train.py]: thanks! sounds like a more natural choice

**Comment by colizz** [comment]
Thanks! Looks like a more natural choice

File: weaver/train.py [hunk 6]
```diff
@@ -778,7 +780,10 @@ def _main(args):
         _logger.warning('Use of `file-fraction` is not recommended in general -- prefer using `data-fraction` instead.')
 
     # training/testing mode
-    training_mode = not args.predict
+    if args.predict:
```
Comment (colizz) [inline weaver/train.py:783]: I also made this change to handle the conflict with `--predict`

**Pull request merged**